src/wp-includes/l10n.php

```diff
@@ -1206,88 +1233,20 @@ function _load_textdomain_just_in_time( $domain ) {
 		return false;
 	}
 
-	$translation_path = _get_path_to_translation( $domain );
-	if ( false === $translation_path ) {
-		return false;
-	}
-
-	return load_textdomain( $domain, $translation_path );
-}
-
-/**
- * Gets the path to a translation file for loading a textdomain just in time.
- *
- * Caches the retrieved results internally.
- *
- * @since 4.7.0
- * @access private
- *
- * @see _load_textdomain_just_in_time()
- *
- * @param string $domain Text domain. Unique identifier for retrieving translated strings.
- * @param bool   $reset  Whether to reset the internal cache. Used by the switch to locale functionality.
- * @return string|false The path to the translation file or false if no translation file was found.
- */
-function _get_path_to_translation( $domain, $reset = false ) {
-	static $available_translations = array();
-
-	if ( true === $reset ) {
-		$available_translations = array();
-	}
-
-	if ( ! isset( $available_translations[ $domain ] ) ) {
-		$available_translations[ $domain ] = _get_path_to_translation_from_lang_dir( $domain );
+	/** @var WP_Textdomain_Registry $wp_textdomain_registry */
+	if ( null === $wp_textdomain_registry->get( $domain ) ) {
+		$wp_textdomain_registry->get_translation_from_lang_dir( $domain );
 	}
 
-	return $available_translations[ $domain ];
-}
-
-/**
- * Gets the path to a translation file in the languages directory for the current locale.
- *
- * Holds a cached list of available .mo files to improve performance.
- *
- * @since 4.7.0
- * @access private
- *
- * @see _get_path_to_translation()
- *
- * @param string $domain Text domain. Unique identifier for retrieving translated strings.
- * @return string|false The path to the translation file or false if no translation file was found.
- */
-function _get_path_to_translation_from_lang_dir( $domain ) {
-	static $cached_mofiles = null;
-
-	if ( null === $cached_mofiles ) {
-		$cached_mofiles = array();
-
-		$locations = array(
-			WP_LANG_DIR . '/plugins',
-			WP_LANG_DIR . '/themes',
-		);
-
-		foreach ( $locations as $location ) {
-			$mofiles = glob( $location . '/*.mo' );
-			if ( $mofiles ) {
-				$cached_mofiles = array_merge( $cached_mofiles, $mofiles );
-			}
-		}
+	$path = $wp_textdomain_registry->get( $domain );
+	if ( ! $path ) {
+		return false;
 	}
 
 	$locale = determine_locale();
-	$mofile = "{$domain}-{$locale}.mo";
+	$mofile = "{$path}/{$domain}-{$locale}.mo"; // TODO: This does not work for themes..
```

Comment:
Does `WP_Textdomain_Registry` need to be aware of whether a domain belongs to a theme or plugin?

Reply:
Not sure, maybe it already helps to check if the path starts with `get_template_directory()` or `get_stylesheet_directory()`. In this case `$mofile` needs to be "{$path}/{$locale}.mo".